Week 4 Day 4 Update: Reliability (#73)
microsoft/win-student-devs: docs/30DaysOfPWA/platforms-practices/04.md

# 4.4: Best Practices for PWA Reliability

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

> Welcome to Week 4, Day 4 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.

Reviewer: ```suggestion
Welcome to Week 4, Day 4 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilities they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.
```

![Placeholder Banner Only. Replace when final assets ready.](_media/day4.png)

### A Quick Note
This article is about the best practices for reliability with PWAs. If you're looking for the basics on offline behavior and caching in PWAs, check out the [Make PWA Work Offline](../core-concepts/05.md) article from Week 1.

## What Does Reliability Look Like For PWA?

> In general, reliability refers to whether or not your application can maintain a consistent user experience under different circumstances. Some of the issus that can follow under the umbrella of reliability:

Reviewer: ```suggestion
In general, reliability refers to whether or not your application can maintain a consistent user experience under different circumstances. Some of the issues that can follow under the umbrella of reliability:
```

* How fast your application loads on varying connection strengths
* How your application works in offline conditions
* How your application and services work as they scale up and increase in usage
* How your application runs on older and less powerful devices

For PWA, reliability issues related to your user's internet connection are particularly of interest, as more traditional web experiences were heavily dependent on the availability of a connection. 

> Today, we're going to build on what we saw about caching in Week 1 to see how we can take advantage of PWA capabilities to make our apps work well regardless of the strength or availability of the user's interent connection.

Reviewer: ```suggestion
Today, we're going to build on what we saw about caching in Week 1 to see how we can take advantage of PWA capabilities to make our apps work well regardless of the strength or availability of the user's internet connection.
```

## Speeding Up Your PWA With Caching Strategies
One of the ways you can speed up your PWA and increase reliability is by minimizing requests to the network and making use of cached responses whenever possible. There are many caching strategies that you can take advantage of, but today we're just going to take a look at two popular options: `cache-first` and `stale-while-revalidate`.

## Cache-First

> In the [Make PWA Work Offline](../core-concepts/05.md) article from Week 1, we saw a little bit about how a cache-first works. A cache-first strategy is exaclty what it sounds like: on a `fetch`, our service will check the cache for the response first, and go to the network _only_ if the cache fails.

Reviewer 2: ```suggestion
In the [Make PWA Work Offline](../core-concepts/05.md) article from Week 1, we saw a little bit about how a cache-first works. A cache-first strategy is exaclty what it sounds like: on a `fetch` event, our service worker will check the cache for the response first, and go to the network _only_ if the cache fails.
```

> In the [Make PWA Work Offline](../core-concepts/05.md) article from Week 1, we saw a little bit about how a cache-first works. A cache-first strategy is exaclty what it sounds like: on a `fetch` event, our service worker will check the cache for the response first, and go to the network _only_ if the cache fails.

Reviewer: ```suggestion
In the [Make PWA Work Offline](../core-concepts/05.md) article from Week 1, we saw a little bit about how a cache-first works. A cache-first strategy is exactly what it sounds like: on a `fetch` event, our service worker will check the cache for the response first, and go to the network _only_ if the cache fails.
```

Let's update our service worker snippet to reflect this:
```javascript
self.addEventListener('fetch', event => {
    event.respondWith((async () => {
        const cache = await caches.open(CACHE_NAME);

        // Try the cache first like last time.
        const cachedResponse = await cache.match(event.request);
        if (cachedResponse !== undefined) {
            // Now, we fetch a new response and cache it in the background
            fetch(event.request).then( response => {
                cache.put(event.request, response.clone());
            });
            // We don't await the above line, so we return our cachedResponse right away
            return cachedResponse;
        } else {
            // Go to the network otherwise
        }
    }
}
```

With this approach, we get the speed benefit of loading a cached response right away, but balance it by keeping our cache relatively up-to-date. Next time we make that request, we will get the updated response instead of the same cached response over and over again.

## Maintaining Data Offline with IndexedDB

> Caching works great for assets, but what about storing structured local data for our? We can use [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) to handle our local data storage to keep our user experience as intact as possible in offline circumstances. IndexedDB allows for asychronous storage of large quantities of structured data that would exceed the capabilites of our cache or LocalStorage. 

Reviewer 2: A good practice for links is to remove the locale path. So in this MDN link, remove the `/en-US` part and leave the rest as is. This way the person's browser will send the necessary request headers for MDN to redirect to their own locale.

Author: I'll double check all these when I go through and convert to aka.

> Caching works great for assets, but what about storing structured local data for our? We can use [IndexedDB](https://aka.ms/learn-pwa/30days-4.4/developer.mozilla.org/docs/Web/API/IndexedDB_API) to handle our local data storage to keep our user experience as intact as possible in offline circumstances. IndexedDB allows for asychronous storage of large quantities of structured data that would exceed the capabilites of our cache or LocalStorage. 

Reviewer: ```suggestion
Caching works great for assets, but what about storing structured local data for our? We can use [IndexedDB](https://aka.ms/learn-pwa/30days-4.4/developer.mozilla.org/docs/Web/API/IndexedDB_API) to handle our local data storage to keep our user experience as intact as possible in offline circumstances. IndexedDB allows for asynchronous storage of large quantities of structured data that would exceed the capabilities of our cache or LocalStorage. 
```

> The API for IndexedDB can be a bit low-level for normal usage, so we'll be taking a look at simple wrapper called LocalForage.

Reviewer 2: How about linking to LocalForage here, since this is the first time it appears in the article?

_Note_: IndexedDB can be used from inside your application and your service worker, choose whatever (or both) works best for your use case.

First, we could create an instance (the same as a database) of localforage and name that instance:

```javascript
// our DB name
const databaseName = "EmailDB"
// create an instance with our name
var emailDatabase = localforage.createInstance({name: databaseName});
```

Now, let's add our data to it using a key-value pair:

```javascript
// Let's get our recent emails as an array
var emails = GetRecentEmails();
// and add that to our database
emailDatabase.setItem("recent-emails", emails);
```

Later, we can fetch our emails using the key and do something with the value:

```javascript
// Asynchronously fetch our emails and then populate our inbox
emailDatabase.getItem("recent-emails").then( emails => {
    PopulateInbox(emails);
});
```
And lastly, we can remove data just as easily:
```javascript
// remove based on key
emailDatabase.removeItem("recent-emails");
```

> Super easy, right? LocalForage keeps its API to just a handful of features for ease of use, and you can learn more about it's capabilities [here.](https://localforage.github.io/localForage/) It's a great option for getting started with IndexedDB and for storing data for simple use cases.

Reviewer 2: ```suggestion
Super easy, right? LocalForage keeps its API to just a handful of features for ease of use, and you can learn more about it's capabilities [here](https://localforage.github.io/localForage/). It's a great option for getting started with IndexedDB and for storing data for simple use cases.
```

> If you need something more flexible for your PWA, there's also [idb](https://github.com/jakearchibald/idb), which is another great IndexedDB wrapper that packs  more of a punch when it comes to functionality. However, it is somewhat more involved when it comes to managing your database and its transactions, so get ready for more complexity.

Reviewer 2: Jake also maintained https://github.com/jakearchibald/idb-keyval which is similar to LocalForage in that it's very very simple to use (and is much smaller than LocalForage). Consider mentioning it too.
